### ISSUE TITLE
`iput --metadata` does not trigger `pep_api_atomic_apply_metadata_operations_post`

### ISSUE THREAD

**Issue by mstfdkmn** [irods/irods#6893]
### iRODS Version, OS and Version
4.3.0 and 4.2.11, almalinux8 and centos7

### What did you try to do?

I wanted to trigger the atomic metadata operations pep by the `--metadata` option of the `iput` command which is I guess the only way in the `iCommands` client to add metadata atomically .

in the core.py file:

```
def pep_api_atomic_apply_metadata_operations_post(rule_args, callback, rei):

    print("I am called!")
```

the client operation:
`iput -f test.txt --metadata "atr; val; unit; atr0; val1; unit1;"`

### Expected behavior

I would expect that the mentioned pep is triggered and it prints out the text of "I am called!" (sends to stdout - logs).

### Observed behavior (including steps to reproduce, if applicable)

`pep_api_atomic_apply_metadata_operations_post ` is not called although the metadata below has been atomically added. Could it be the case that `--metadata` is not supported by the atomic metadata api?
```
[irods@dev-u0137480-instance1 irods]$ imeta ls -d test.txt
AVUs defined for dataObj /u0137480_instance1/home/rods/test.txt:
attribute:  atr0
value:  val1
units:  unit1
----
attribute: atr
value:  val
units:  unit

```

**Comment by alanking** [comment]
I think this is due to the fact that `iput` invokes the put API and that is the only API PEP which will be triggered as a result of the `iput`, at least on the connected server. Any further API calls made from within the iRODS agent will not normally trigger any API PEPs.

This commit changed the behavior of `iput --metadata` to use the atomic metadata API: https://github.com/irods/irods/commit/9042130cad8300bae6871e1f5038c0262f5d05ab However, the PEP will not be triggered because the API call is happening inside the iRODS agent.

As @korydraughn is (maybe? :p) suggesting, the API PEP could be triggered if a redirection occurs from a catalog consumer to the provider as it would need to make a "client" connection from the consumer to the provider and make the API call.

**Comment by korydraughn** [comment]
Correct.

**Comment by korydraughn** [comment]
With just a provider, the atomic metadata PEPs will not fire.

The atomic metadata API is invoked via the server-side call. That means unless there's a redirect to another server, no PEPs will be triggered for that API call.

**Comment by trel** [comment]
So, is there anything to do here?   Provide an alternate PEP for @mstfdkmn to watch/instrument?

**Comment by korydraughn** [comment]
There's nothing to do here in regard to the atomic metadata PEP.

@mstfdkmn You need to detect the PUT PEP and inspect the `condInput` (i.e. the `KeyValPair`) member of the `DataObjInp` object.

I believe the key you're looking for is `"metadataIncluded"` which is defined below:
https://github.com/irods/irods/blob/03e7e0515ab3e367c969ddc783b0e0d3dd995924/lib/core/include/irods/rodsKeyWdDef.h#L33

That key will be mapped to the AVU string you passed on the command line.

**Comment by korydraughn** [comment]
`metadataIncluded` will be a key in the `condInput.key` list. Its value will share the same index as the key.

For example, in C, if `metadataIncluded` is at `condInput.key[5]`. Then the AVU string will be at `condInput.value[5]`. The same thing will apply to the PREP.

Looking at the PREP code, it looks like you can do the following:
```python
def pep_api_data_obj_put_post(rule_args, callback, rei):
    print(rule_args[2].condInput['metadataIncluded'])
```

**Comment by mstfdkmn** [comment]
This is what I needed. However could you confirm that you meant [this](https://github.com/irods/irods_rule_engine_plugin_python/blob/4b1567fcef8047c5519559c5f544cd58f2497c56/irods_types.cpp#L374) by

> Looking at the PREP code

? Thanks.

**Comment by korydraughn** [comment]
That is the correct file, but wrong section.

I looked at the following:
- https://github.com/irods/irods_rule_engine_plugin_python/blob/4b1567fcef8047c5519559c5f544cd58f2497c56/irods_types.cpp#L212-L231

**Comment by mstfdkmn** [comment]
Actually I would like to learn how I can know `condInput` has the key `metadataIncluded`. I am currently sruggling with peps - not only this one but other peps have `condInput` in their `rule_args[2]`. For example I would like to get `dataId` with pep_api_data_obj_unlink_post, I inspected `condInput` and cannot find `dataId` there.

**Comment by trel** [comment]
What Kory said (save in PRE, use in POST)... because after it's unlinked... it's too late.

https://docs.irods.org/4.3.0/system_overview/tips_and_tricks/#using-temporarystorage-in-the-irods-rule-language

**Comment by mstfdkmn** [comment]
Yes, I have already used the pre-pep not only this but for other peps as well...Sorry not part of this issue, but could you try to find `dataId` in `pep_api_data_obj_unlink_pre`?

I could only find a solution like the one here:
```
def pep_api_data_obj_unlink_pre(rule_args, callback, rei):
    coll_name, data_name = os.path.split(str(rule_args[2].objPath))
    query = Query(callback, 'DATA_ID', f"COLL_NAME = '{coll_name}' and DATA_NAME = '{data_name}'",
                  case_sensitive=True, output = AS_LIST)
    global data_id
    data_id = [row[0] for row in query][0]
    print(data_id)
```

I cannot find always a desired value in chained keys.

**Comment by korydraughn** [comment]
> but could you try to find `dataId` in `pep_api_data_obj_unlink_pre`?

Yes, that is exactly what we're recommending that you do.

> I cannot find always a desired value in chained keys.

It's not clear to me what you mean here. Please elaborate.

**Comment by mstfdkmn** [comment]
I meant I could not find `dataId` in `pep_api_data_obj_unlink_pre` like the my initial try below, and asked you whether you can find it for me? 
```
def pep_api_data_obj_unlink_pre(rule_args, callback, rei):
    global data_id
    data_id = rule_args[2].condInput['dataId'])
```

`irods stdout | {"log_category":"legacy","log_facility":"local0","log_level":"error","log_message":"caught python exception: Traceback (most recent call last):\n  File \"/etc/irods/core.py\", line 11, in pep_api_data_obj_unlink_pre\n    data_test = str(rule_args[2].condInput['dataId'])\nKeyError: 'dataId was not found in the list of keys.'\n","request_api_name":"DATA_OBJ_UNLINK_AN","request_api_number":615,"request_api_version":"d","request_client_user":"rods","request_host":"10.0.2.100","request_proxy_user":"rods","request_release_version":"rods4.3.0","server_host":"dev-u0137480-instance1.irods.t.icts.kuleuven.be","server_pid":87268,"server_timestamp":"2023-02-10T13:33:54.432Z","server_type":"agent"}`

Basically your recommendation is that if cannot find what you need in the serialized args of peps, then to use a query and assign result of that query to a global var. Thanks.

**Comment by trel** [comment]
Correct - the particular items available will be different for every operation - and in this case, unlink's input is a logical path, not a data_id.

**Comment by korydraughn** [comment]
Exactly what @trel said.

Is it safe to say this issue can be closed?

**Comment by trel** [comment]
I vote yes - but let's wait for @mstfdkmn.

**Comment by mstfdkmn** [comment]
Yes I guess this can be closed as long as you dont see anything with the original part of the issue - with just a provider, the atomic metadata PEPs will not fire -. Btw thanks for the answers given to extensions here!

**Comment by trel** [comment]
Thank you.
